test/unit/Cardano/WalletLayerSpec.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.WalletLayerSpec
+    ( spec
+    ) where
+
+
+import Prelude
+
+import Cardano.DBLayer
+    ( DBLayer (..), PrimaryKey (..) )
+import Cardano.DBLayer.MVar
+    ( newDBLayer )
+import Cardano.NetworkLayer.HttpBridgeSpec
+    ( mockNetworkLayer, noLog )
+import Cardano.Wallet
+    ( WalletId (..), WalletName (..) )
+import Cardano.Wallet.AddressDerivationSpec
+    ()
+import Cardano.Wallet.AddressDiscovery
+    ( SeqState )
+import Cardano.Wallet.AddressDiscoverySpec
+    ()
+import Cardano.Wallet.MnemonicSpec
+    ()
+import Cardano.Wallet.Primitive
+    ( SlotId (..) )
+import Cardano.WalletLayer
+    ( NewWallet (..), WalletLayer (..), mkWalletLayer )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( runExceptT )
+import Data.Either
+    ( isLeft, isRight )
+import Data.Maybe
+    ( isJust )
+import Test.Hspec
+    ( Spec, describe, it, shouldSatisfy )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, checkCoverage )
+import Test.QuickCheck.Monadic
+    ( monadicIO )
+
+import qualified Data.Text as T
+
+spec :: Spec
+spec = do
+    describe "WalletLayer works as expected" $ do
+        it "Wallet upon creation is written down in db and cannot be created more than once"
+            (checkCoverage walletCreationProp)
+        it "Wallet after being created can be got using valid wallet Id"
+            (checkCoverage walletGetProp)
+        it "Wallet with wrong wallet Id cannot be got"
+            (checkCoverage walletGetWrongIdProp)
+
+
+{-------------------------------------------------------------------------------
+                                    Test Logic
+-------------------------------------------------------------------------------}
+
+data WalletLayerFixture = WalletLayerFixture {
+      _fixtureDBLayer :: DBLayer IO SeqState
+    , _fixtureWalletLayer :: WalletLayer IO SeqState
+    }
+
+prepareWalletLayerFixture :: IO WalletLayerFixture
```

Comment:
Now that we are designing tests from start - do we really want to do the same test design as we did before? Defining fixtures like this, which promote generating data within function body (instead of generating fixture and other arguments as function arguments) breaks quickchecks shrink functionality.

Maybe we should reconsider this approach and generate everything within function arguments so that property body is as deterministic as possible. In this specific example everything works correctly as `mockNetworkLayer` is pure deterministic function. But I am afraid that this design encourages of someone in future to add
```haskell
  somethingArbitrary <- pick arbitrary
```

and to add `somethingArbitrary` to `WalletLayerFixture` to use later which would break quickchecks shrink functionality.

Maybe this is a good oportunity to have a chat with Edsko about https://iohk.io/blog/an-in-depth-look-at-quickcheck-state-machine/

Reply:
I feel that the current fixture and test setup is very simple. `setupFixture` prepares db, network, walletLayer and creates wallet returning its db, walletLayer and corresponding walletId. For every test, the same fixture is instantiated. No complex transitions are occurring here. Maybe address  `state machine` concept in separate PR and refactor then?

Reply:
@akegalj Your remark is sound and, note that we aren't going for the same approach as before. Here, the arbitrary generation happens via QuickCheck in the functions arguments. The fixture doesn't generate anything (and I agree, it OUGHT NOT), but rather, setup a state given some already generated data.

It will be quite tricky to do differently as, some I/O needs to happen for the property to happen, and the property is intrinsically monadic. It would be wrong to shove in any `pick` or `generate` call in the property. Yet, as they stand, the properties are totally deterministic (modulo the DB failure). 

Beside, state machine testing is another stage in the testing process. They won't replace those basic unit-tests, but rather come as a complement. The nice thing with the DBLayer as it stands right now makes it possible to test the "MVar DBLayer" well, and then, use it as a model for state machine testing, comparing a more complex implementation (like an SQLite DBLayer) to the simple MVar one.

Reply:
:+1: